extensions/s3/src/main/java/io/deephaven/extensions/s3/S3ChannelContext.java

```diff
@@ -36,6 +35,11 @@ final class S3ChannelContext extends BaseSeekableChannelContext implements Seeka
     final S3AsyncClient client;
     final S3Instructions instructions;
 
+    /**
+     * Used to temporarily hold a request to prevent it from being GC'd.
+     */
+    S3Request.AcquiredRequest acquiredRequest;
```

Comment:
This can be a local variable.

Reply:
In my previous patch, I made it a local variable and then the compiler optimized it out. I have explained it here: https://github.com/deephaven/deephaven-core/pull/5613#discussion_r1638945647
Then I moved it to a member field.

Reply:
Got it; that makes sense. If we move the fill method to AcquiredRead though, it won't be able to optimize it away, and it could become a local variable again.

Reply:
Generally, when the compiler optimizes something away and you workaround it in some fashion, I worry that while the compiler doesn't optimize it away today, it _could_ optimize it away in the future (now, maybe there are JVM specs that more clearly define cases that it will never optimize away, but that's pretty bespoke; and at a very minimum, would deserve a big comment around the variable that "this can't be made local"). In these situations, I prefer to take a step back and try to re-architect the dependencies so that we actually use the hard-reference in some fashion.

Reply:
So I couldn't move the implementation of `fill` method inside AcquiredRequest because it touches a lot of local files from inside `Request`. What I did was just added an accessor for `fill` in AcquiredRequest and made the `Request.fill` a private implementation.

After that, I converted the class field to a local variable in this commit and was still seeing CancellationException, which means the underlying buffer got freed while we were filling the first request, meaning the compiler optimized it out too. This is the commit: https://github.com/deephaven/deephaven-core/pull/5613/commits/9af2bf145bcd69a35b321b69b4c778b1f4e62a1d
The exception I got
```
Caused by: java.io.IOException: Cancelled an operation while fetching fragment ix=25 [524288000, 545259519]/20971520 dh-nbc-s3-test1/parquet_65536/file_00019_20240517130133.gz.parquet
	at io.deephaven.extensions.s3.S3ChannelContext.handleS3Exception(S3ChannelContext.java:207)
	at io.deephaven.extensions.s3.S3Request.fillImpl(S3Request.java:222)
	at io.deephaven.extensions.s3.S3Request$AcquiredRequest.fill(S3Request.java:71)
	at io.deephaven.extensions.s3.S3ChannelContext.fill(S3ChannelContext.java:125)
	at io.deephaven.extensions.s3.S3SeekableByteChannel.read(S3SeekableByteChannel.java:88)
	... 27 more

Caused by: java.util.concurrent.CancellationException
	at java.base/java.util.concurrent.CompletableFuture.cancel(CompletableFuture.java:2510)
	at io.deephaven.extensions.s3.S3Request.cleanup(S3Request.java:241)
	at io.deephaven.util.reference.CleanupReferenceProcessor$DrainQueue.run(CleanupReferenceProcessor.java:156)
	... 1 more
```

So I reverted back to a class field and added a comment.

Reply:
Pushed another PR merging with the ideas that Devin had suggested: https://github.com/devinrsmith/deephaven-core/tree/sm-s3-cache-fix